File: glox/scanner/scanner.go
package scanner

import (
	"fmt"

	"gsmayya.com/glox/utils"
)

var start int = 0
var current int = 0
var line int = 1

func setValues() {
	start = 0
	current = 0
	line = 1
}

func ScanTokens(source string) []Token {
	// start of new scanning, so set values
	setValues()
	fmt.Println("Scanning : " + source)
	var tokens []Token
	for !isAtEnd(source) {
		start = current
		scanToken(source, &tokens)
	}
	tokens = append(tokens, Token{Type: EOF, Lexeme: "", Literal: nil, Line: line})
	return tokens
}

func scanToken(source string, tokens *[]Token) {
	var c byte = advance(source)
	switch c {
	case '(':
		addToken(source, tokens, LEFT_PAREN)
	case ')':
		addToken(source, tokens, RIGHT_PAREN)
	case '{':
		addToken(source, tokens, LEFT_BRACE)
	case '}':
		addToken(source, tokens, RIGHT_BRACE)
	case ',':
		addToken(source, tokens, COMMA)
	case '.':
		addToken(source, tokens, DOT)
	case '-':
		addToken(source, tokens, MINUS)
	case '+':
		addToken(source, tokens, PLUS)
	case ';':
		addToken(source, tokens, SEMICOLON)
	case '*':
		addToken(source, tokens, STAR)
	case '!':
		if match(source, '=') {
			addToken(source, tokens, BANG_EQUAL)
		} else {
			addToken(source, tokens, BANG)
		}
	case '=':
		if match(source, '=') {
			addToken(source, tokens, EQUAL_EQUAL)
		} else {
			addToken(source, tokens, EQUAL)
		}
	case '<':
		if match(source, '=') {
			addToken(source, tokens, LESS_EQUAL)
		} else {
			addToken(source, tokens, LESS)
		}
	case '>':
		if match(source, '=') {
			addToken(source, tokens, GREATER_EQUAL)
		} else {
			addToken(source, tokens, GREATER)
		}
	case '/':
		if match(source, '/') {
			// A comment goes until the end of the line.
			for peek(source) != '\n' && !isAtEnd(source) {
				advance(source)
			}
		} else {
			addToken(source, tokens, SLASH)
		}
	case ' ', '\r', '\t':
		// Ignore whitespace.
	case '\n':
		line++
	case '"':
		handleString(source, tokens)
	default:
		if isDigit(c) {
			number(source, tokens)
			break
		}
		if isAlpha(c) {
			identifer(source, tokens)
			break
		}
		utils.Error(line, "Unexpected character.")
	}
}

func match(source string, expected byte) bool {
	if isAtEnd(source) {
		return false
	}
	if source[current] != expected {
		return false
	}
	current++
	return true
}

func peek(source string) byte {
	if isAtEnd(source) {
		return '\000'
	}
	return source[current]
}

func advance(source string) byte {
	c := source[current]
	current++
	return c
}

func addToken(source string, tokens *[]Token, tType TokenType) {
	addTokenWithLiteral(source, tokens, tType, nil)
}

func addTokenWithLiteral(source string, tokens *[]Token, tType TokenType, literal any) {
	text := source[start:current]
	*tokens = append(*tokens, Token{Type: tType, Lexeme: text, Literal: literal, Line: line})
}

func isAtEnd(source string) bool {
	return current >= len(source)
}

func handleString(source string, tokens *[]Token) {
	for peek(source) != '"' && !isAtEnd(source) {
		if peek(source) == '\n' {
			line++
		}
		advance(source)
	}

	if isAtEnd(source) {
		utils.Error(line, "Unterminated string.")
		return
	}

	// The closing ".
	advance(source)

	// Trim the surrounding quotes.
	value := source[start+1 : current-1]
	addTokenWithLiteral(source, tokens, STRING, value)
}

func isDigit(c byte) bool {
	return c >= '0' && c <= '9'
}

func number(source string, tokens *[]Token) {
	for isDigit(peek(source)) {
		advance(source)
	}

	// Look for a fractional part.
	if peek(source) == '.' && isDigit(peekNext(source)) {
		// Consume the "."
		advance(source)

		for isDigit(peek(source)) {
			advance(source)
		}
	}

	value := source[start:current]
	addTokenWithLiteral(source, tokens, NUMBER, value)
}

func peekNext(source string) byte {
	if current+1 >= len(source) {
		return '\000'
	}
	return source[current+1]
}

func isAlpha(c byte) bool {
	return (c >= 'a' && c <= 'z') ||
		(c >= 'A' && c <= 'Z') ||
		c == '_'
}

func isAlphaNumeric(c byte) bool {
	return isAlpha(c) || isDigit(c)
}

func identifer(source string, tokens *[]Token) {
	for isAlphaNumeric(peek(source)) {
		advance(source)
	}
	text := source[start:current]
	tType, ok := keywords[text]
	if !ok {
		tType = IDENTIFIER
	}
	addToken(source, tokens, tType)
}

File: jlox/app/src/main/java/com/craftinginterpreters/lox/Scanner.java
package com.craftinginterpreters.lox;

import static com.craftinginterpreters.lox.TokenType.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
/**
 * Scans the texts and generates tokens for the text. 
 * String of text == Tokens in the Grammar. 
 * Parser will get these tokens and generates the meaning ful. 
 */
public class Scanner {

    /**
     * Set of keywords
     */
    private static final Map<String, TokenType> keywords;

    /**
     * Statically assign them. 
     * TODO: Can we move this along with tokentype to ensure that checking of tokens is easier? 
     */
    static {
        keywords = new HashMap<>();
        keywords.put("and", AND);
        keywords.put("class", CLASS);
        keywords.put("else", ELSE);
        keywords.put("false", FALSE);
        keywords.put("for", FOR);
        keywords.put("fun", FUN);
        keywords.put("if", IF);
        keywords.put("nil", NIL);
        keywords.put("or", OR);
        keywords.put("print", PRINT);
        keywords.put("return", RETURN);
        keywords.put("super", SUPER);
        keywords.put("this", THIS);
        keywords.put("true", TRUE);
        keywords.put("var", VAR);
        keywords.put("while", WHILE);
    }

    // Source string. 
    private final String source;
    // Generated tokens. 
    private final List<Token> tokens = new ArrayList<>();

    /** 
     * Variables for state maintance during scanning. 
     * Start pointer 
     * Current pointer 
     * Number of lines
     */
    private int start = 0;
    private int current = 0;
    private int line = 1;

    /**
     * Constructor with the string
     */
    Scanner(String source) {
        this.source = source;
    }

    /**
     * Main driver. 
     * While the source does not end, 
     * scan for each token, 
     * At the end EOF -- Special token 
     */
    List<Token> scanTokens() {
        while (!isAtEnd()) {
            start = current;
            scanToken();
        }

        tokens.add(new Token(EOF, "", null, line));
        return tokens;
    }

    /**
     * State maintenance to see if the characters in source string has ended. 
     */
    private boolean isAtEnd() {
        return current >= source.length();
    }

    /**
     * Core engine for scanning. Goes through each character and based on that, creates token 
     * For multi-character it peeks ahead to get that token 
     * Comments are handled by checking and ignoring until end. 
     * on default, it can be either 
     *  Number -> Scanned set of digits 
     *  identifier -> Already declared or a token known. 
     */
    private void scanToken() {
        char c = advance();
        switch (c) {
            case '(':
                addToken(LEFT_PAREN);
                break;
            case ')':
                addToken(RIGHT_PAREN);
                break;
            case '{':
                addToken(LEFT_BRACE);
                break;
            case '}':
                addToken(RIGHT_BRACE);
                break;
            case ',':
                addToken(COMMA);
                break;
            case '.':
                addToken(DOT);
                break;
            case '-':
                addToken(MINUS);
                break;
            case '+':
                addToken(PLUS);
                break;
            case ';':
                addToken(SEMICOLON);
                break;
            case '*':
                addToken(STAR);
                break;
            case '!':
                addToken(match('=') ? BANG_EQUAL : BANG);
                break;
            case '=':
                addToken(match('=') ? EQUAL_EQUAL : EQUAL);
                break;
            case '<':
                addToken(match('=') ? LESS_EQUAL : LESS);
                break;
            case '>':
                addToken(match('=') ? GREATER_EQUAL : GREATER);
                break;
            case '/':
                if (match('/')) {
                    // a comment goes until the end of the line so ignore all
                    while (peek() != '\n' && !isAtEnd())
                        advance();
                } else {
                    addToken(SLASH);
                }
                break;
            case ' ':
            case '\r':
            case '\t':
                // ignore the white
                break;
            case '\n':
                line++;
                break;
            case '"':
                string();
                break;
            default:
                if (isDigit(c)) {
                    number();
                } else if (isAlpha(c)) {
                    identifier();
                } else {
                    Lox.error(line, "Unexpected character.");
                }
        }
    }

    /**
     * Identifier -> 
     *   Find if the subset string is a Token 
     *   If not, then it is identifier and add that as in. 
     */
    private void identifier() {
        while (isAlphaNumeric(peek()))
            advance();
        String text = source.substring(start, current);
        TokenType type = keywords.get(text);
        if (type == null)
            type = IDENTIFIER;
        addToken(type);
    }

    /** 
     * Simple check for alphabets. 
     */
    private boolean isAlpha(char c) {
        return (c >= 'a' && c <= 'z') ||
                (c >= 'A' && c <= 'Z') ||
                c == '_';
    }

    /**
     * For alphanumeric
     */
    private boolean isAlphaNumeric(char c) {
        return isAlpha(c) || isDigit(c);
    }

    /**
     * Numeric checks
     */
    private boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    /**
     * Number -> 
     *   Integer -> Only digits 
     *   Floating point -> with a decimal 
     */
    private void number() {
        while (isDigit(peek()))
            advance();

        // Look for the fractional part, ensure that there is digits after '.' 
        if (peek() == '.' && isDigit(peekNext())) {
            // consume the "."
            advance();

            while (isDigit(peek()))
                advance();
        }

        addToken(NUMBER, Double.parseDouble(source.substring(start, current)));
    }

    /**
     * Peeks Next to current pointer and gets a character, do not advance "current" pointer
     */
    private char peekNext() {
        if (current + 1 >= source.length())
            return '\0';
        return source.charAt((current + 1));
    }

    /** 
     * Quoted strings. Consumes anything between quotes '"'
    */
    private void string() {
        while (peek() != '"' && !isAtEnd()) {
            if (peek() == '\n')
                line++;
            advance();
        }

        if (isAtEnd()) {
            Lox.error(line, "Unterminated String");
        }
        // the closing "
        advance();

        // Trim the surrounding quotes.
        String value = source.substring(start + 1, current - 1);
        addToken(STRING, value);
    }

    /**
     * Matches with "advance" -> Moves the pointer while consuming the character and matching with expected. 
     * This is used to handle multi-character tokens like !=
     */
    private boolean match(char expected) {
        if (isAtEnd())
            return false;
        if (source.charAt(current) != expected)
            return false;
        current++;
        return true;
    }

    /**
     * Peeks the current pointer without "advance"
     */
    private char peek() {
        if (isAtEnd())
            return '\0';
        return source.charAt(current);
    }

    /**
     * Moves the current pointer, "advance" and returns the character at the current pointer. 
     */
    private char advance() {
        return source.charAt(current++);
    }

    /**
     * If identified as Token, then add it. Without any literal Object
     */
    private void addToken(TokenType type) {
        addToken(type, null);
    }

    /**
     * Adds token with the value assigned, like NUMBER (actual Number). STRING (actual string)
     * For known tokens the literal will be null
     */
    private void addToken(TokenType type, Object literal) {
        String text = source.substring(start, current);
        tokens.add(new Token(type, text, literal, line));
    }

}
